src/swap/do_action.rs

```diff
@@ -0,0 +1,212 @@
+use crate::{swap::db, SwapId};
+use comit::Timestamp;
+
+#[async_trait::async_trait]
+pub trait Do<E>
+where
+    Self: CheckMemory<E> + ShouldAbort + Execute<E> + Remember<E>,
+    E: Clone + Send + Sync + 'static,
+    <Self as Execute<E>>::Args: Send + Sync,
+{
+    async fn r#do(
+        &self,
+        beta_expiry: Timestamp,
+        execution_args: <Self as Execute<E>>::Args,
+    ) -> anyhow::Result<Next<E>> {
+        if let Some(event) = self.check_memory().await? {
+            return Ok(Next::Continue(event));
+        }
+
+        if self.should_abort(beta_expiry).await? {
+            return Ok(Next::Abort);
+        }
+
+        let event = Execute::<E>::execute(self, execution_args).await?;
+        self.remember(event.clone()).await?;
+
+        Ok(Next::Continue(event))
+    }
+}
+
+#[async_trait::async_trait]
+impl<E, A> Do<E> for A
+where
+    A: CheckMemory<E> + ShouldAbort + Execute<E> + Remember<E>,
+    E: Clone + Send + Sync + 'static,
+    <Self as Execute<E>>::Args: Send + Sync,
+{
+}
+
+#[async_trait::async_trait]
+pub trait CheckMemory<E> {
+    async fn check_memory(&self) -> anyhow::Result<Option<E>>;
+}
+
+#[async_trait::async_trait]
+impl<E, A> CheckMemory<E> for A
+where
+    A: db::Load<E> + std::ops::Deref<Target = SwapId>,
+    E: 'static,
+{
+    async fn check_memory(&self) -> anyhow::Result<Option<E>> {
+        self.load(**self).await
+    }
+}
+
+#[async_trait::async_trait]
+pub trait ShouldAbort {
+    async fn should_abort(&self, beta_expiry: Timestamp) -> anyhow::Result<bool>;
+}
+
+#[async_trait::async_trait]
+impl<A> ShouldAbort for A
+where
+    A: BetaLedgerTime + Sync,
+{
+    async fn should_abort(&self, beta_expiry: Timestamp) -> anyhow::Result<bool> {
+        let beta_ledger_time = self.beta_ledger_time().await?;
+
+        Ok(beta_expiry <= beta_ledger_time)
+    }
+}
+
```

Comment:
This could just be a function which gets rid of the trait altogether :)

```suggestion
async fn should_abort<A>(actor: &A, beta_expiry: Timestamp) -> anyhow::Result<bool> where A: BetaLedgerTime + Sync {
    let beta_ledger_time = self.beta_ledger_time().await?;

    Ok(beta_expiry <= beta_ledger_time)
}
```

Reply:
Generally, be mindful when designing things around the notion of "instances" or "objects". Often times, a function does the job equally well without adding more abstraction layers :)

Reply:
In fact, you can even get rid of the 2nd trait as-well, by just having a function that takes both timestamps and compares them :)

What is the different you may ask?

It pays off to split code into "decision logic" that just depends on data and "integration logic" that wires different modules together. Decision logic that just depends on data is easy to test and "integration logic" that doesn't make any decisions (because they are delegated to decision modules) but just wires them together is hard to do wrong in Rust.

Reply:
I would be happy to split the decision logic into a function, because this happens to be the very conservative decision logic of Nectar with regards to taking actions or not. As you say, it would make sense for it to take two timestamps and to not know about where they come from, and it would be easier to test.

I would keep the trait though because it allows implementations to define their own logic.